fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -1075,10 +1092,15 @@ public RecordMetaData build() {
      */
     @Nonnull
     public RecordMetaData build(boolean validate) {
+        for (Index index : indexes.values()) {
+            if (isIndexCounterBasedSubspaceKeyEnabled() && !index.subspaceKeyIsExplicitlySet()) {
+                index.setSubspaceKey(String.valueOf(indexSubspaceKeyCounter++));
```

Comment:
> This shouldn't be a string--it should just be the integer.

It was initially integer. But I noticed that at decode time it becomes long, which made it hard to compare equality. But I see your concern, especially with the clash. This should be fixed.

> I also think this should probably go in `addIndex` rather than `build`

Assigning them in addIndex makes it harder, as we will need to deal with duplicates too. Also it could result in unpredictable behavior depending on when the counter-based approach is enabled.

> Then the first two indexes will be added with subspace keys that use the old default whereas the third index uses the new (counter-based) scheme. 

I think I disagree with this to be a benefit. It complicates the scheme, and it's also at odds with the non-default behavior that you had in-mind earlier in `RecordMetaData.build` and your next comment about unstable allocation. If this counter-based scheme is enabled, it should be enabled for all of the subspace keys that are not explicitly set and it should not rely on the order of these calls. With the current scheme, you can still set the subspace key to name, by explicitly calling `setSubspaceKey(name)`.

> The other benefit is I think the current counter allocation code is unstable.

Sure, I agree. And we can fix it with TreeMap.

> If it used a more stable map (like a `TreeMap`), it would still have a problem with removed indexes.

No I don't think so. `indexSubspaceKeyCounter` is persisted in the proto for the exact same reason and existing indexes have explicit subspace keys, so nothing will implicitly change.

Reply:
The kind of stability I'm getting at is when the user *doesn't* use a meta-data proto and instead relies on the stability of mutations to the `RecordMetaDataBuilder` to maintain their meta-data. 

> > I also think this should probably go in `addIndex` rather than `build`
> 
> Assigning them in addIndex makes it harder, as we will need to deal with duplicates too. Also it could result in unpredictable behavior depending on when the counter-based approach is enabled.

Maybe I don't understand what unpredictable behavior you're worried about. I think with my suggestion, any index added to the meta-data prior to enabling the counter continues to use whatever key it was using before. Any index added to the meta-data after to enabling the counter gets a value determined by when it was added. We already depend on index addition order to make sure that the `added` version of indexes is stable/accurate (as well as the `removed` versions of `FormerIndex`es). I think this is more of the same.

> > Then the first two indexes will be added with subspace keys that use the old default whereas the third index uses the new (counter-based) scheme.
> 
> I think I disagree with this to be a benefit. It complicates the scheme, and it's also at odds with the non-default behavior that you had in-mind earlier in `RecordMetaData.build` and your next comment about unstable allocation.

I agree this is different from my suggestion in `RecordMetaData.build` if there are record types missing explicit subspace keys insofar as (one of) my suggestions was just to throw an error. But the fundamental invariant is that index subspace keys need to be stable across software upgrades or record stores will get into a really corrupt state. This suggestion here is essentially to allow the (sophisticated) user a way to enable to the new behavior on new indexes but keep the old default on existing indexes (by starting to use the counter half-way through).

I don't think I follow how this would result in unstable allocation of subspace keys.

> If this counter-based scheme is enabled, it should be enabled for all of the subspace keys that are not explicitly set and it should not rely on the order of these calls. With the current scheme, you can still set the subspace key to name, by explicitly calling `setSubspaceKey(name)`.
> 
> > The other benefit is I think the current counter allocation code is unstable.
> 
> Sure, I agree. And we can fix it with TreeMap.
> 
> > If it used a more stable map (like a `TreeMap`), it would still have a problem with removed indexes.
> 
> No I don't think so. `indexSubspaceKeyCounter` is persisted in the proto for the exact same reason and existing indexes have explicit subspace keys, so nothing will implicitly change.

If the user initially has this:

```java
metaDataBuilder.addIndex("MyRecord", index1);
metaDataBuilder.addIndex("MyRecord", index2);
```

And changes their code to start doing this:

```java
metaDataBuilder.addIndex("MyRecord", index1);
metaDataBuilder.addIndex("MyRecord", index2);
metaDataBuilder.removeIndex(index1.getName());
```

Then (assuming that `index1.getName() < index2.getName()`) the index subspace key for `index2` will change and also the former index corresponding to `index1` will just be straight up wrong.

Another example would be something like if the user initially did:

```java
metaDataBuilder.addIndex("MyRecord", index1);
metaDataBuilder.addIndex("MyRecord", index3);
```

So `index1` is assigned key 0 and `index3` assigned key `1`. Then they update their code to:


```java
metaDataBuilder.addIndex("MyRecord", index1);
metaDataBuilder.addIndex("MyRecord", index3);
metaDataBuilder.addIndex("MyRecord", index2);
```

So `index1` is assigned key 0 and `index2` is assigned key 1 and `index3` is assigned key 2. After upgrading the meta-data, it will look to the user like all data from `index3` has disappeared.

Reply:
I suppose another approach that avoids this kettle of fish would be to disallow using this feature without initializing the builder from a meta-data proto file. I'm not sure how I feel about that.

Reply:
Alec and I talked offline. I was not aware that by contract, users are not allowed to change order of `addIndex` / `removeIndex`. We decided the following: 

- `addIndex` will assign the subspace keys.
- at build time, we'll validate that the keys are unique. 

If we are starting from a proto file, the subspace keys will always start from 1. And, since the contract is the order will not change, the subspace keys will remain stable.